### PR TITLE
[4.0] [plg_system_actionlogs] Convert to prepared statement

### PR DESCRIPTION
**Pull request by HLeithner**
# Summary of Changes
Updated SQL queries to prepared statements and made some cleanups around the queries.

### Testing Instructions
Use the plugin in all ways you can think of.

Use the privacy component and trigger action log entries.

### Expected result
Nothing changed.

### CODE REVIEW

File: plugins/system/actionlogs/actionlogs.php [hunk 1]
```diff
@@ -174,14 +175,18 @@ public function onContentPrepareData($context, $data)
...
-			->from($this->db->quoteName('#__action_logs_users'))
-			->where($this->db->quoteName('user_id') . ' = ' . (int) $data->id);
+		$db = $this->db;
+		$id = (int) $data->id;
+
+		$query = $db->getQuery(true)
+			->select($db->quoteName(['notify', 'extensions']))
+			->from($db->quoteName('#__action_logs_users'))
+			->where($db->quoteName('user_id') . ' = :userid')
+			-bind(':userid', $id, ParameterType::INTEGER);
```
Comment (Quy) [inline plugins/system/actionlogs/actionlogs.php]:
```suggestion
			->bind(':userid', $id, ParameterType::INTEGER);
```

File: plugins/system/actionlogs/actionlogs.php [hunk 2]
```diff
@@ -364,65 +374,84 @@ public function onUserAfterSave($user, $isNew, $success, $msg)
...
+			$columns = ['user_id', 'notify'];
+
+			$query->bind($values, [$userid, $notify], ParameterType::INTEGER);
 
 			if (isset($user['actionlogs']['actionlogsExtensions']))
 			{
-				$values[]  = $this->db->quote(json_encode($user['actionlogs']['actionlogsExtensions']));
+				$values[]  = ':extension';
 				$columns[] = 'extensions';
+				$extension = json_encode($user['actionlogs']['actionlogsExtensions'];
```
Comment (Quy) [inline plugins/system/actionlogs/actionlogs.php]:
```suggestion
				$extension = json_encode($user['actionlogs']['actionlogsExtensions']);
```

File: plugins/system/actionlogs/actionlogs.php [hunk 3]
```diff
@@ -364,65 +374,84 @@ public function onUserAfterSave($user, $isNew, $success, $msg)
...
+			$values = [$db->quoteName('notify') . ' = :notify'];
+
+			$query->bind(':notify', $notify, ParameterType::INTEGER);
 
 			if (isset($user['actionlogs']['actionlogsExtensions']))
 			{
-				$values[] = $this->db->quoteName('extensions') . ' = ' . $this->db->quote(json_encode($user['actionlogs']['actionlogsExtensions']));
+				$values[] = $db->quoteName('extensions') . ' = :extension';
+				$extension = json_encode($user['actionlogs']['actionlogsExtensions']);
+				$query->bind(':extension', $extension)
```
Comment (Quy) [inline plugins/system/actionlogs/actionlogs.php]:
```suggestion
				$query->bind(':extension', $extension);
```

File: plugins/system/actionlogs/actionlogs.php [hunk 4]
```diff
@@ -364,65 +374,84 @@ public function onUserAfterSave($user, $isNew, $success, $msg)
...
 		// If preferences don't exist, insert.
 		if (!$exists && $authorised && isset($user['actionlogs']))
 		{
-			$values  = array((int) $user['id'], (int) $user['actionlogs']['actionlogsNotify']);
-			$columns = array('user_id', 'notify');
+			$notify  = (int) $user['actionlogs']['actionlogsNotify'];
+			$values  = [':userid', ':notify'];
+			$columns = ['user_id', 'notify'];
+
+			$query->bind($values, [$userid, $notify], ParameterType::INTEGER);
```
Comment (SharkyKZ) [inline plugins/system/actionlogs/actionlogs.php]: `[$userid, $notify]` has to be a variable.

File: plugins/system/actionlogs/actionlogs.php [hunk 5]
```diff
@@ -282,10 +289,13 @@ public function onAfterRespond()
...
 		if ($daysToDeleteAfter > 0)
 		{
-			$conditions = array($db->quoteName('log_date') . ' < ' . $query->dateAdd($now, -1 * $daysToDeleteAfter, ' DAY'));
+			$days = -1 * $daysToDeleteAfter;
 
 			$query->clear()
-				->delete($db->quoteName('#__action_logs'))->where($conditions);
+				->delete($db->quoteName('#__action_logs'))
+				->where($db->quoteName('log_date') . ' < ' . $query->dateAdd(':now', ':days', 'DAY'))
+				->bind([':now', ':days'], [$now, $days], [ParameterType::STRING, ParameterType::INTEGER]);
```
Comment (SharkyKZ) [inline plugins/system/actionlogs/actionlogs.php]: Use separate `bind()` calls or assign `[$now, $days]` to a variable.

And this is failing on PostgreSQL:

>SQLSTATE[HY093]: Invalid parameter number: :days

Comment (HLeithner) [reply plugins/system/actionlogs/actionlogs.php]: Fixed. thx

**Comment by waader** [comment]
I have tested this item :white_check_mark: successfully on 4472095fce489e3f341f8f01a20ac55aec3d08f3<br /><br />Tested with postgresql.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25503">issues.joomla.org/tracker/joomla-cms/25503</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f2b26e56b659933bf56139483b20ff52afad1284<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25503">issues.joomla.org/tracker/joomla-cms/25503</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/25503">issues.joomla.org/tracker/joomla-cms/25503</a>.</sub>

**Pull request merged**